python/dgl/contrib/multi_gpu_datastore.py

```diff
@@ -0,0 +1,156 @@
+##
+#   Copyright 2021 Contributors 
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+""" The MultiGPUDataStore class. """
+
+from .. import backend as F
+
+class MultiGPUDataStore:
+    """ Class for storing a large tensor split across GPU memory according to
+    nodes. When reading memory from oehr GPUs, the call must be from all
+    GPUs storing the data.
+
+    Example:
+    --------
+
+    It can be created via the shape and data type information from the tensor
+    to split across gpus `large_cpu_tensor` in the below code:
+
+    >>> split_data = MultiGPUTensor(large_cpu_tensor.shape,
+    ...                               large_cpu_tensor.dtype,
+    ...                               dev_id,
+    ...                               nccl_comm)
+    >>> split_data.set_global(large_cpu_tensor)
+
+    Then, once it is stored across GPU memory, during training, feature for
+    mini-batches can be fetched via the `get_global()` method. If we have the
+    tensor of mini-batch nodes `input_nodes` we can use the following code to
+    fetch the features for the mini-batch:
+    ...
+    >>> batch_features = split_data.all_gather_row(input_nodes)
```

Comment:
Do we need to call the `split_data.all_gather_row(input_nodes)` from all the processes simultaneously?

Reply:
Yes, the call cannot return until all processes have called it.